### PR TITLE
add block split bloom filter

### PR DESCRIPTION
**Pull request by kangpinghuang**
1. use block split bloom filter
2. add distinct value to bloom filter
3. add ordinal index to bloom filter index

Issue: #1389

### CODE REVIEW

File: be/src/olap/rowset/segment_v2/bloom_filter_index_reader.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    ColumnBlock block(_reader->type_info(), (uint8_t*)&value, &nullmap, num_to_read, _pool.get());
+    ColumnBlockView column_block_view(&block);
+
+    RETURN_IF_ERROR(_bloom_filter_iter.seek_to_ordinal(ordinal));
+    size_t num_read = num_to_read;
+    RETURN_IF_ERROR(_bloom_filter_iter.next_batch(&num_read, &column_block_view));
+    DCHECK(num_to_read == num_read);
+    // construct bloom filter
+    BloomFilter::create(_reader->_bloom_filter_index_meta.algorithm(), bf);
+    auto ret = (*bf)->init(value.data, value.size, _reader->_bloom_filter_index_meta.hash_strategy());
```
Comment (imay) [inline be/src/olap/rowset/segment_v2/bloom_filter_index_reader.cpp]: I'm not sure why you can read bloom fiter from _bloom_filter_iter. I see that you what you write into IndexedColumnWriter is not bloom filter binary.

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/olap/olap_cond.cpp [hunk 2]
```diff
@@ -459,7 +459,52 @@ bool Cond::eval(const BloomFilter& bf) const {
...
+bool Cond::eval(const segment_v2::BloomFilter* bf) const {
+    //通过单列上BloomFilter对block进行过滤。
+    switch (op) {
+    case OP_EQ: {
+        bool existed = false;
+        if (operand_field->is_string_type()) {
+            Slice* slice = (Slice*)(operand_field->ptr());
+            existed = bf->test_bytes(slice->data, slice->size);
+        } else {
+            int32_t* v = (int32_t*)operand_field->ptr();
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/olap_cond.cpp]: ok， I will remove it.

File: be/src/olap/rowset/segment_v2/block_split_bloom_filter.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+namespace doris {
+namespace segment_v2 {
+
+const uint32_t BlockSplitBloomFilter::SALT[8] = {0x47b6137b, 0x44974d91, 0x8824ad5b, 0xa2b7289d,
+        0x705495c7, 0x2df1424b, 0x9efc4947, 0x5c6bfb31};
+
+void BlockSplitBloomFilter::add_hash(uint64_t hash) {
+    // most significant 32 bit mod block size as block index(BTW:block size is
+    // power of 2)
+    uint32_t block_index = (uint32_t)(hash >> 32) & (_size / BYTES_PER_BLOCK - 1);
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/block_split_bloom_filter.cpp]: I will add a judgement to check whether the _size is power of 2.

File: be/src/olap/rowset/segment_v2/block_split_bloom_filter.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    uint32_t key = (uint32_t)hash;
+
+    // Calculate masks for bucket.
+    uint32_t masks[8];
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/block_split_bloom_filter.cpp]: ok

File: be/src/olap/rowset/segment_v2/encoding_info.cpp [hunk 5]
```diff
@@ -180,39 +193,48 @@ EncodingInfoResolver::EncodingInfoResolver() {
     _add_map<OLAP_FIELD_TYPE_TINYINT, BIT_SHUFFLE>();
     _add_map<OLAP_FIELD_TYPE_TINYINT, FOR_ENCODING, true>();
     _add_map<OLAP_FIELD_TYPE_TINYINT, PLAIN_ENCODING>();
+    _add_map<OLAP_FIELD_TYPE_TINYINT, BLOOM_FILTER>();
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/encoding_info.cpp]: remove it

File: gensrc/proto/segment_v2.proto [hunk 6]
```diff
@@ -54,6 +54,7 @@ enum EncodingTypePB {
     DICT_ENCODING = 5;
     BIT_SHUFFLE = 6;
     FOR_ENCODING = 7; // Frame-Of-Reference
+    BLOOM_FILTER = 8;
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply gensrc/proto/segment_v2.proto]: remove it

File: be/src/olap/utils.h [hunk 7]
```diff
@@ -512,6 +512,10 @@ bool valid_datetime(const std::string& value_str);
         }\
     } while (0)
 
+// get the highest bit of 1 in uint32_t
+// copy from java Integer.highestOneBit()
+uint32_t highest_one_bit(uint32_t i);
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/utils.h]: ok

File: be/src/olap/rowset/segment_v2/bloom_filter_page.h [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+                // value exists, just skip
+                continue;
+            }
+            if (_is_slice_type()) {
+                Slice* value = (Slice*)p;
+                _bf->add_bytes(value->data, value->size);
+            } else {
+                _bf->add_bytes((char*)p, (uint32_t)sizeof(CppType));
+            }
+            _distinct_values.insert(*p);
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/bloom_filter_page.h]: I will fix it

File: be/src/olap/rowset/segment_v2/bloom_filter_page.h [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+template<>
+struct BloomFilterTraits<Slice> {
+    using ValueDict = std::set<Slice, Slice::Comparator>;
+};
+
+// BloomFilterPage is for building bloom filter of data
+//
+template<FieldType field_type>
+class BloomFilterPageBuilder : public PageBuilder {
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/bloom_filter_page.h]: I move this logic to BloomFilterIndexWriter

File: be/src/olap/rowset/segment_v2/bloom_filter_index_reader.cpp (hunk 1 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/bloom_filter_index_reader.cpp]: I change the logic to build bloom filter agaist the data page.

File: be/src/olap/rowset/segment_v2/options.h [hunk 10]
```diff
@@ -28,6 +30,12 @@ struct PageBuilderOptions {
     size_t data_page_size = DEFAULT_PAGE_SIZE;
 
     size_t dict_page_size = DEFAULT_PAGE_SIZE;
+
+    HashStrategyPB hash_strategy = HASH_MURMUR3_X64_64;
```
[the remark this answers is not in the document]
Comment (kangpinghuang) [reply be/src/olap/rowset/segment_v2/options.h]: remove it

**Pull request closed without merging**